File: Longest Cycle in a Graph.cpp
class Solution {
public:
    void dfs(int u,vector<bool>& vis,vector<bool>& dfsvis,vector<int>& count,vector<int>& edges,int& ans){
        
        static int time = 0;
        time+=1;
        count[u] = time;
        vis[u] = true;
        dfsvis[u] = true;
        int v = edges[u];
        if(v == -1){
            dfsvis[u] = false;
            return;
        }
            
            if(vis[v] and dfsvis[v]){
                ans = max(ans,count[u]-count[v]);
                dfsvis[u] = false;
                return;
            }else if(!vis[v]){
                dfs(v,vis,dfsvis,count,edges,ans);
            }
        dfsvis[u] = false;
    }
    int longestCycle(vector<int>& edges) {
        int ans = -1;
        int n=edges.size();
        vector<bool> vis(n,false);
        vector<bool> dfsvis(n,false);
        vector<int> count(n,0);
        
        for(int i=0;i<n;i++){
            if(!vis[i] and edges[i] != -1){
                dfs(i,vis,dfsvis,count,edges,ans);
            }
        }
        if(ans == -1)
            return -1;
        
        return ans+1;
    }
};

File: Longest Cycle in a Graph.java
class Solution {
    int max=-1;
    public int longestCycle(int[] edges) {
        boolean vis[]=new boolean[edges.length];
       
        for(int i:edges)
        {  if(i==-1||vis[i]) continue;
           HashMap<Integer,Integer> map=new HashMap<>();
           dfs(i,0,vis,map,edges);
        }
        return max;
    }
    public void dfs(int node,int dis,boolean[]vis, HashMap<Integer,Integer> map,int []edges)
    {
        if(node==-1) return ;
        if(vis[node])
        {   
            if(!map.containsKey(node)) return ;
            max=Math.max(dis-map.get(node),max);
            //System.out.println(max);
            return ;
        }
        vis[node]=true;
        map.put(node,dis);
        dfs(edges[node],dis+1,vis,map,edges);
        return ;
         
    }
}
